### DISCUSSION TITLE
BiblioteQ Mac install question, driver not loaded.

### DISCUSSION THREAD

**Discussion by KeviZol** [textbrowser/biblioteq#349] [Q&A] [answered]
Hi, I try to install BiblioteQ 2023.12.30 from latest d.dmg file on Sonoma 14.3.1. macOS. I copy the app file over to Applications folder, I remove App Translocation in Terminal with xattr, and I copy my custom biblioteq.conf file to same Applications folder. Both files have same owner (me) and groups (admin). When trying to connect to my PostgreSQL DB, app says 'Unable to open a database connection with the provided information'. However with same conf file I could connect successfully on a W11 Biblioteq app install. Error log says a driver is not loaded related to Source/biblioteq_c.cc file, line number 2833. User guide says in some cases conf file shall be copied to /, however I could not copy that over either with sudo cp -p command. What am I missing here? I read BQ app could still look for the dmb conf file in some cases, and since it is either not available or does not include DB details, connection attempt drops with error. Could anyone help out kindly? Thanks in advance. ZK

**Comment by textbrowser** [comment]
I don't include the PQ driver because it involves much work in maintaining the dependencies in Homebrew. Best that you build BQ on your own if you desire the splendi-o-sity of a real database engine. See you.

**Comment by textbrowser** [comment]
The Windows version includes PQ I think. Lots of work for such a little task, right? I agree. Much work. I tried automating it, however, that's also a lot of work because Homebrew is not consistent with Intel and Apple Silicon. Separate paths. Then there is the Qt project file. A little bit there and more bits there and then every little bit becomes a huge bit. So I don't do it for the release because it may remain inconsistent for others. I simply build BQ on my MacOS and it works because my Homebrew and local project file are correct.

**Comment by textbrowser** [comment]
You don't need to install it in Applications. I have it on my Desktop. BQ doesn't require special privileges.

**Comment by KeviZol** [comment]
Hi, thank you for such quick reply! 😊
I would say I am ’superuser’ level, I can tweak, change and more or less understand tech instructions, but could build a package (have not done so fow 30 or so years…).
I use BQ for the family library, and use PQ on a Synology NAS so that family members (and I) could reach DB from different locations.
I read and understand you are overwhelmed, just wonder if you ouwld have any further enthusiasm to make a full build (with PQ driver) like once a year or so roughly.
For minimal individual users I would guess having the very latest features would not be of utmost priority, rather to be able to follow changes like yearly with some simle update steps (schema, application, etc.).

Would you remember perhaps what is the latest Mac BQ  build including PQ? I remember I could use the dmg build some time back.

Thank you kindly for you time reading this.

**Comment by KeviZol** [comment]
Yes, Windows version includes PQ luckily, as otherwise I would lose DB reach of many days/weeks of DB polishing…
It is just that not all family members have Windows, but MacOS.
I understand compilation is like a hairpulling activity, it seems no change in that from 30 years ago… 😊

Thank you for confirming it is not my wrong doing of package ’installation’ on Mac, if you had the mood once here and there to compile a full fledge release, many mini users would be grateful I think.

But I know time is money…

Have a nice evening!

**Comment by KeviZol** [comment]
Yes, thank you, I think the wall I bounced to is that the recent dmg bundles do not handle PQ DBs, 'only' local SQLite.
Thank you kindly for all your answers and clarifications!
Zoltan

**Comment by textbrowser** [comment]
BQ compiles easily as I've made most of its dependencies optional. Well, on Debian, let's say, I expect that the person performing the compilation is not a cat. Some experience is required, especially if they're on Debian or FreeBSD. Beyond that, and perhaps Windows, it's easy. YAZ is kind of replaced by SRU. PQ requires libpq and libpq requires TLS, etc.

I'm going to provide an Apple Silicon release because I will not have Intel MacOS.

**Comment by KeviZol** [comment]
Yes, Debian and FreeBSD are tricky and need some experience...
Fort he Mac however, I could easily just copy and paste the app and conf files earlier and worked lieka charm in 20 seconds as you say easy...
It was when the build included the PQ driver, so I will now check the latest build with PQ included and deploy that, hopefully it will handle latest schema fine.

I will keep my Intel MacOS for a while.

Thanks again for all the inputs!
Cheers

**Comment by textbrowser** [comment]
Schema-related revisions are referenced in the release notes.

https://github.com/textbrowser/biblioteq/blob/master/SQL/postgresql_update_schema.sql

That includes 2023. Older releases were removed and are available in GH history logs, somewhere.

Will include Intel and Silicon releases for MacOS as long as possible. Silicon is preferred so the Intel version will disappear in the future.

**Comment by KeviZol** [comment]
Thank you for further details.
Interestingly, 2022.05.30 build works on High Sierra, buti t has this PQ issue on Sonoma 14.3.1.
Anyway, thank you kindly for your kind quick answers really! Perhaps a PQ included build might help some strugglers like I am…

Have a nice week! 😊
Zoltan

**Comment by KeviZol** [comment]
Hi again,

Really sorry for bothering you again, I just tried several dmg packages, but with no success in connecting to the PQ DB.
Would you recall perhaps what was the last dmg MacOS package to which you included the PostgreSQL drivers as well?

I would try to carry on with that version.

Thank you for your attention!

Best regards,
Zoltan

From: Zoltan Keviczky ***@***.***>
Sent: 20 February 2024 00:14
To: textbrowser/biblioteq ***@***.***>
Subject: RE: [textbrowser/biblioteq] BiblioteQ Mac install question, driver not loaded. (Discussion #349)

Thank you for further details.
Interestingly, 2022.05.30 build works on High Sierra, buti t has this PQ issue on Sonoma 14.3.1.
Anyway, thank you kindly for your kind quick answers really! Perhaps a PQ included build might help some strugglers like I am…

Have a nice week! 😊
Zoltan

**Reply by textbrowser** [reply]
I don't and if I did I would not recall it because if there was a subsequent release requiring a database revision, the recommended version per your request would not work.

**Comment by KeviZol** [comment]
Thank you for timely reply, yes that might be a risk however I have a 2011 iMac with 20190530 package still working fine (have not added any items for some time now though...).
Unfotunately that package is not on the list anymore, but you are right it might not work with latest DB schema.
Have a nice day!

**Comment by textbrowser** [comment]
I built the newest version with the newest Qt on an Intel. The install target which is part of the generated make file copies many required libraries. However, some libraries are not copied and these require that I revise the Qt project file to include them and possibly execute otool with them. This may be a problem on Apple Silicon too. Qt warns about this in their macdeployqt documentation. And this is the reason I do not include the PQ libraries. It's a lot of work because Homebrew's structure includes version information in directory names. Whenever a library's version is changed, so is its directory structure. And so, I need to modify the project file.

**Comment by textbrowser** [comment]
https://doc.qt.io/qt-6/macos-deployment.html

As can be seen, a lengthy document inspires a complicated process. This is the work involved in releasing BQ and PQ on MacOS. It requires that I spend many hours in an initial release and one or many hours afterwards. And there is zero guarantee that it will work beyond my computer. So, why?

**Comment by KeviZol** [comment]
Oh, I can see it is quite a hectic series of steps… Thank you really for bethering with the question!
So what would you suggest?
I use BQ to manage, track and control our wider family (parents,2 brothers) library with I think cca 2-3k books and growing. So far I have added some hundreds only manually.
I chose BQ as it knows PQ and as such I can keep its DB on my NAS with relative ease (steps described on the internet at super user level like I am, and I do not like to keep my things in external clouds…, and family members can reach DB and I can make DB dumps, etc.). I preferred not to use local DB (eg. SQLite) as it would be too vulnerable to any local problems, like PC crashes, etc.
I myself use the Windows BQ version (exe installs) with PQ seemlessly.
Do you tend to keep PQ included tot he Windows versions and make exe from time to time, like every other year or so?
Would you suggest the trying to get some Windows emulation to Mac PCs which some of my family members use?
Shall I/Can I carry on adding more books to PQ DB? Or I would risk losing the timely built DB if and when PQ option would disppear?

These might be too many questions, apologies really, just would be keen to know how to plan, and intention is really NOT to bother you.
And to tell the truth, there is not really any alternatives, for the above purpose BQ is by far the best! 😊

Thank you kindly for your attention!

Best of regards,
Zoltan

**Comment by textbrowser** [comment]
Patience.

The Windows build requires diligence too. Only the Unix (non-MacOS) are simple because the dependencies are easier.

**Comment by KeviZol** [comment]
Sure thing, I got it, so i will keep on adding the books tot he DB.
Thank you for your kind attention to my case. I bet you have plenty...
Best regards,
Zoltan

**Comment by textbrowser** [comment]
I've been working on the PQ driver almost every day.

Qt 5 via Homebrew does not produce good results because the Qt driver is not included in Homebrew's Qt tree. SQLite driver is.

Qt 6 via Homebrew does not include all of the required dependencies (set aside /usr/local/ Cellar) after BiblioteQ is created. I cannot detect which dependency fails because the driver is loaded when the connection is attempted.

Qt via the online installer refers to an /Applications libpq dependency. This would require that I create libpq or copy Homebrew's. I tried Homebrew's libpq and this creates a cascade of dependencies.

**Comment by textbrowser** [comment]
And assuming that I have it work for Intel, it will not work for Apple Silicon because Homebrew paths are different and so the Qt project file would need to consider both Intel and Silicon.

I think the best solution is have people install Homebrew's Qt.

**Comment by KeviZol** [comment]
Hi, thank you a ton for considering and investigating the use case! Sad to hear Qt puts such complicated dependencies. ☹

I have Homebrew Qt (installed with the Terminal line ’brew install qt’, no additional plugins, just the dependencies which Homebrew auto installed along with Qt), and the laptop is Intel based Mac, yet the 20231230 d.dmg release still does not find the DB driver according to BQ error log.
Is there any plugin I am still missing from Qt perhaps?

Kind regards,
Zoltan

**Comment by textbrowser** [comment]
Right. It would not because that version of BQ was built with the official Qt and not the Homebrew Qt. Let me prepare a new BQ with Homebrew, upload it, and see if it works for you.

**Comment by KeviZol** [comment]
Sura thing, that is very kind of you! Thank you!

**Comment by textbrowser** [comment]
You'll need to install qt-postgresql.

**Comment by textbrowser** [comment]
```
saturn@cosmos ~ % brew info qt-postgresql
==> qt-postgresql: stable 6.6.2 (bottled)
Qt SQL Database Driver
https://www.qt.io/
/usr/local/Cellar/qt-postgresql/6.6.2 (9 files, 156.2KB) *
  Poured from bottle using the formulae.brew.sh API on 2024-03-24 at 12:51:53
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/q/qt-postgresql.rb
License: GPL-2.0-only or GPL-3.0-only or LGPL-3.0-only
==> Dependencies
Build: cmake ✘
Required: libpq ✔, qt ✔
==> Analytics
install: 155 (30 days), 462 (90 days), 1,637 (365 days)
install-on-request: 155 (30 days), 462 (90 days), 1,637 (365 days)
build-error: 0 (30 days)
```

**Comment by textbrowser** [comment]
https://github.com/textbrowser/biblioteq/releases/edit/2024.03.25

Will be temporary and removed. Download quick.

**Comment by KeviZol** [comment]
Thank you, I have it installed. Now I have both qt and qt-postgresql installed via Homebrew. No clashes reported (so far).

**Comment by textbrowser** [comment]
I'll remove in 10 minutes from now.

**Comment by KeviZol** [comment]
I have downloaded it, you can remove...

**Comment by textbrowser** [comment]
OK. Let us all know of the results.

**Comment by textbrowser** [comment]
Did you connect to your PQ database?

**Comment by KeviZol** [comment]
Sorry to say, it still says database error/driver not loaded… ☹

I did full uninstall of previous version I tried with (20231230), installed the one you just shared over GutHub, updated config file tom y DB. All these after successfully installing qt-postgresql via Homebrew…

PS: I need to rush for my younger kid now, I will be back at desk in cca 45 minutes…

**Comment by textbrowser** [comment]
Very peculiar. Works here and not there. I'll check again.

**Comment by KeviZol** [comment]
Yes, strange, i turned my VPN off, firewall off, still... On Parallels Windows 11 with the Windows instance it connects fine, even with firewall and VPN (as it should).
I will also check if I missed a step. Perhaps will uninstall both qt and qt-posgresql, and intall only the latter in Homebrew...

**Comment by textbrowser** [comment]
Will try in 3 or so minutes.

**Comment by KeviZol** [comment]
With same instance?

**Comment by textbrowser** [comment]
Never mind. Qt 6 Homebrew does not work. That's some bullshit. I will see about Qt 6 official. I know one works.

**Comment by KeviZol** [comment]
I could install an earlier version of ***@***.******@***.***>. Just would need to know the x.x it works with...
I will be offline for a while, but checking the emails now and then.
Thank you really for bothering! Perhaps a Qt5 based qt-postgresql version could solve the case short term...

**Comment by textbrowser** [comment]
OK. Have something.

**Comment by textbrowser** [comment]
The official Qt plus the Homebrew libpq are OK here.

**Comment by textbrowser** [comment]
Uploading now.

**Comment by textbrowser** [comment, marked as answer]
brew install libpq

(You'll have this already but I need a reminder for future people or something.)

https://formulae.brew.sh/formula/libpq

**Comment by textbrowser** [comment]
OK. Uploaded. Give that a try.

**Comment by textbrowser** [comment]
You must have the newest libpq: /usr/local/Cellar/libpq/16.2_1/lib/libpq.dylib.

**Comment by KeviZol** [comment]
OK, i have downloaded it, will give it a try now and get back to you.

**Comment by textbrowser** [comment]
It works right? Yup, it must work.

**Comment by textbrowser** [comment]
Remember that you must have the newest Homebrew libpq which is the one I mentioned up there.

**Comment by textbrowser** [comment]
If this works, it will complicate things because of the specific PQ version. And for Silicon, the path is different.

**Comment by textbrowser** [comment]
But that is easily corrected with an exists() test.

**Comment by textbrowser** [comment]
Did you connect?

**Comment by KeviZol** [comment]
I have that, just checked it now, intalling now and will get back to you in a min

**Comment by KeviZol** [comment]
Wow, i think it works! Yuppeee! 😊
For now it says I shall update my schema to 20240325, just as I used so far only the 20231230 version and schema.
But this means it reaches out and checks and works! 😊
Wow Wow Wow! THank you, thank you! 😊

**Reply by textbrowser** [reply]
You'll have seen a reminder that you need to upgrade your schema. It's a basic reminder that doesn't inspect every statement for a particular release so it can be bypassed but if it is that's your fault.

**Comment by KeviZol** [comment]
I might need some help here (as well), would not know how to correct...

**Comment by textbrowser** [comment]
To correct the schema? You need to execute the statements in the included SQL file.

**Comment by KeviZol** [comment]
With qt-postgresql install libpq is also installed. It just needs to be checked before all future updates, or set to auto updaze in Homebrew...
Could you share perhaps the schema query of latest 20240325 package, so that I could update it?

**Comment by KeviZol** [comment]
Oh yes, sorry, I did things in a rush, sure, just doing in a moment...

**Comment by textbrowser** [comment]
https://github.com/textbrowser/biblioteq/blob/master/SQL/postgresql_update_schema.sql

**Comment by textbrowser** [comment]
```brew upgrade``` suffices.

**Comment by textbrowser** [comment]
As for the BQ schema, I erased much of the stuff in the SQL file and retained 2023 and future.

**Comment by textbrowser** [comment]
I update schemas through the psql command, executing the statements in the SQL file mentioned.

**Comment by textbrowser** [comment]
The SQLite portion of BQ includes a nicer feature which is not included for PQ because PQ requires administrator knowledge and that is required for PQ and BQ.

**Comment by KeviZol** [comment]
Just updated the schema, and it works like a charm. Reaches the DB, I could open an item. Superb! 😊
I need to check the Windows versoin as well now, as I think it is only the 20231230 version as the latest one.

**Comment by textbrowser** [comment]
The Windows version I have not released and will not work because of the new fields. You can always build it yourself.

**Comment by KeviZol** [comment]
I can imagine, but I think you cannot use SQLite approach, if you want to have the DB in your own NAS/Server, allowing some family memebrs to be able to reach the same DB, and also I myself would use Mac OSX or Windows 10/11 instance, depending where I am at...
Could I use SQLite for this busines case? I think I will have to remain with PQ...

**Comment by KeviZol** [comment]
Sure, got it, I will use the Mac OS version then until the new Windows release… I just opened the Windows version, and it popluated the items, I guess adding new items would be a problem…
But that is amazing you could fix the Mac version! 😊 Thank you very much!
Am not a rich guy, happy to send to you 30-40 EURs for beer… 😊

**Comment by KeviZol** [comment]
Yes, so far I used the 20231230 schema, as that is the Windows version I used so far.
From now on I will use the Mac Os version for DB management.
Going forward, I would not think this Qt install steps would be of an issue to other dummies like I am, so perhaps this can be a way forward for same Mac OS dummies like me...
I do not know the exists command you proposed, but will check, as I plan to keep my Intel based laptop for a while for sure...

**Comment by textbrowser** [comment]
The Windows version will work because the new database fields are not required by your BQ version. The exists() function is a Qt-ism.

**Comment by textbrowser** [comment]
https://github.com/textbrowser/biblioteq/commit/463430a4c0d4c24c87a72adb11cbe6910515ad5a

**Comment by KeviZol** [comment]
OK, thank you, I thought so about the exists(), so will check it in Qt docs...

**Comment by KeviZol** [comment]
Case closed! 😊
Thank you again a whole lot! 😊
